certbot/plugins/manual.py

```diff
@@ -59,203 +52,113 @@ class Authenticator(common.Plugin):
+        self._validate_hooks()
+
+    def _validate_hooks(self):
+        if self.config.validate_hooks:
+            for hook_prefix in ('auth', 'cleanup',):
+                hook = self.conf('{0}-hook'.format(hook_prefix))
+                if hook is not None:
+                    hooks.validate_hook(hook, '--{0}'.format(hook_prefix))
 
     def more_info(self):  # pylint: disable=missing-docstring,no-self-use
-        return ("This plugin requires user's manual intervention in setting "
-                "up challenges to prove control of a domain and does not need "
-                "to be run as a privileged process. When solving "
-                "http-01 challenges, the user is responsible for setting up "
-                "an HTTP server. Alternatively, instructions are shown on how "
-                "to use Python's built-in HTTP server. The user is "
-                "responsible for configuration of a domain's DNS when solving "
-                "dns-01 challenges. The type of challenges used can be "
-                "controlled through the --preferred-challenges flag.")
+        return (
+            'This plugin allows the user to customize setup for domain '
+            'validation challenges either through shell scripts provided by '
+            'the user or by performing the setup manually.')
 
     def get_chall_pref(self, domain):
         # pylint: disable=missing-docstring,no-self-use,unused-argument
         return [challenges.HTTP01, challenges.DNS01]
 
-    def perform(self, achalls):
-        # pylint: disable=missing-docstring
-        self._get_ip_logging_permission()
-        mapping = {"http-01": self._perform_http01_challenge,
-                   "dns-01": self._perform_dns01_challenge}
+    def perform(self, achalls):  # pylint: disable=missing-docstring
+        self._verify_ip_logging_ok()
+
+        if self.conf('auth-hook'):
+            perform = self._perform_achall_with_script
+        else:
+            perform = self._perform_achall_manually
+
         responses = []
-        # TODO: group achalls by the same socket.gethostbyname(_ex)
-        # and prompt only once per server (one "echo -n" per domain)
         for achall in achalls:
-            responses.append(mapping[achall.typ](achall))
+            perform(achall)
+            responses.append(achall.response(achall.account_key))
         return responses
 
-    @classmethod
-    def _test_mode_busy_wait(cls, port):
-        while True:
-            time.sleep(1)
-            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-            try:
-                sock.connect(("localhost", port))
-            except socket.error:  # pragma: no cover
-                pass
+    def _verify_ip_logging_ok(self):
+        if not self.conf('public-ip-logging-ok'):
+            cli_flag = '--{0}'.format(self.option_name('public-ip-logging-ok'))
+            msg = ('NOTE: The IP of this machine will be publicly logged as '
```

Comment:
I did this because I prefer single quotes for strings, however, the lines using double quotes include single quote characters. Rather than escaping them, I just used double quotes which is consistent with [our](https://certbot.eff.org/docs/contributing.html#coding-style) [style](https://www.python.org/dev/peps/pep-0008/#string-quotes) [guides](https://google.github.io/styleguide/pyguide.html?showone=Strings#Strings).